text/0000-programmatic-style-injection.md

```diff
@@ -0,0 +1,172 @@
+---
+title: Programmatic style injection
+status: DRAFTED
+created_at: 2021-09-16
+updated_at: 2021-09-30
+pr: (leave this empty until the PR is created)
+---
+
+# Programmatic style injection
+
+## Summary
+
+This proposal describes how LWC components can programmatically associate CSS stylesheets to an LWC component.
+
+## Basic example
+
+```js
+import { LightningElement } from 'lwc';
+import myStyles from './styles.css';
+
+export default Example class extends LightningElement {
+    static stylesheets = [
+        myStyles
+    ];
+}
+```
+
+## Motivation
+
+CSS custom properties and CSS `::part()` are the standard mechanisms by which applications can influence the look and feel of shadow DOM components. Component authors are required to explicitly define those extension points that can be styled from outside the shadow tree. While this approach gives a certain level of control over the component appearance, in certain cases developers might want to take full control over the styles injected in the shadow tree. And this without compromising the component encapsulation. 
+
+Component libraries are a prime example of this use case. Component authors want to expose high-quality base components with minimal styles and let application developers add their branding.
+
+LWC implicit template and stylesheet resolution system makes it impossible today to extend from a base LWC component and override its styles at the same time. This proposal addresses this issue by exposing a mechanism to associate stylesheets directly with the LWC component class. 
+
+## Detailed design
+
+A new `stylesheets` static property is added on the `LightningElement` constructor. This property allows developers to define which stylesheets the component should use. It accepts an array of [LWC stylesheets](#lwc-stylesheets-modules). The base `LightningElement.stylesheets` static property returns an empty array.
+
+The `stylesheets` property works on both light DOM and shadow DOM LWC components. This stylesheet injection mechanism complements the existing implicit template/stylesheet resolution mechanism.
+
+The LWC engine caches the static `stylesheets` value for the lifetime of the application during the component class definition. Modifying the `stylesheets` property after the component code is evaluated will not affect the stylesheet injected by this component. 

```

Comment:
@pmdartus It's unclear to me whether dynamically modifying the `stylesheets` _array_ is supported or not:

```js
this.constructor.stylesheets = [foo] // definitely not supported

this.constructor.stylesheets.push(foo) // supported?
```

I'd rather go ahead and say that the entire array is frozen (i.e. throws an error on modification).

However if we say this, then I guarantee some people are going to want to dynamically modify their stylesheets on-demand. 🙂 And sadly people have found a (hacky) way to do this, which (AFAICT) this proposal does not cover.

Reply:
Thinking back to [how component authors are able to dynamically inject stylesheets today using a workaround](https://github.com/salesforce/lwc-test/pull/147), it occurs to me that the current RFC does not actually solve this use case, because it's not possible to modify stylesheets dynamically during the component lifecycle. So component authors will still be tempted to use the workaround.

I am beginning to wonder if it makes sense to have a slightly different design. Something like:

```js
render() {
  return {
    template: template,
    stylesheets: [stylesheetX, stylesheetY]
  }
}
```

In the above case, the `render()` function could return either 1) a template, as is done today, or 2) an object with a `template` property and and optional `stylesheets` property.

In the above design, we would not need to specify stylesheet ordering, because it would work exactly the same as the current `render()` function works today. If a stylesheet is not injected, it is injected on-demand.

OTOH, it would not solve the problem of multi-template components that render A, then B, then A. (The stylesheets from B would still have been injected after A, so B will take precedence.) To fix that, we would need to overhaul how stylesheet injection is done (either [remove stylesheets as we attempted once](https://github.com/salesforce/lwc/pull/2797), or bump stylesheets later in the list when they are rendered).

Reply:
I like the approach. The only downside with this is that it requires importing the template manually.